Switch sigtest maven plugin

liweinan opened the pull request.

liweinan commented:
With this configuration the build in [jberet-tck-runner](https://github.com/jberet/jberet-tck-runner) generates this error:

```
[ERROR] Failed to execute goal jakarta.tck:sigtest-maven-plugin:2.2:generate (default-cli) on project com.ibm.jbatch.tck.sigtest.exec: Point <classes>to-directory-with-classfiles-to-test</classes> in plugin config section! -> [Help 1]
```

---

<img width="3008" alt="image" src="https://github.com/jberet/jberet-tck-porting/assets/201907/b8d3a598-9918-435a-9e24-32d0d01c4396">

liweinan commented:
@jamezp It seems I didn't configure the `sigtest` plugin correctly. Could you please help to review this when you have time? Thanks!

jamezp commented on src/main/resources/runners/sigtest/pom.xml:
```diff
@@ -127,22 +128,32 @@
+                        <!--                    <phase>integration-test</phase>-->
+                        <goals>
+                            <!--                      <goal>check</goal>-->
+                            <goal>generate</goal>
```
This needs to be `check`.

jamezp commented on src/main/resources/runners/sigtest/pom.xml:
```diff
@@ -17,82 +17,83 @@
-                            <outputDirectory>${project.build.directory}/sigtest-copy</outputDirectory>
-                        </configuration>
-                    </execution>
+                    <!--                    <execution>-->
+                    <!--                        <id>unpack-sigfiles</id>-->
+                    <!--                        <goals>-->
+                    <!--                            <goal>unpack</goal>-->
+                    <!--                        </goals>-->
+                    <!--                        <phase>pre-integration-test</phase>-->
+                    <!--                        <configuration>-->
+                    <!--                            <artifactItems>-->
+                    <!--                                <artifactItem>-->
+                    <!--                                  <groupId>jakarta.batch</groupId>-->
+                    <!--                                  <artifactId>com.ibm.jbatch.tck</artifactId>-->
+                    <!--                                  <version>${project.parent.version}</version>-->
+                    <!--                                </artifactItem>-->
+                    <!--                            </artifactItems>-->
+                    <!--                            <includes>sigtest/*</includes>-->
+                    <!--                            <outputDirectory>${project.build.directory}/sigtest-copy</outputDirectory>-->
+                    <!--                        </configuration>-->
+                    <!--                    </execution>-->
```
You will still need this.

jamezp commented on src/main/resources/runners/sigtest/pom.xml:
```diff
@@ -127,22 +128,32 @@
-                  <failOnError>true</failOnError>
-                  <packages>jakarta.batch.**</packages>
-                  <classes>${project.build.directory}/dependency</classes>
+                    <sigfile>${project.build.directory}/jakarta.batch.sig_${project.parent.version}</sigfile>
```
This value should be `${project.build.directory}/sigtest-copy/sigtest/sigtest-1.6-batch.standalone.tck.sig-2.1-se11-OpenJDK-J9`.

jamezp commented on src/main/resources/runners/sigtest/pom.xml:
```diff
@@ -127,22 +128,32 @@
+                    <packages>
+                        jakarta.batch
+                    </packages>
+                    <attach>false</attach>
```
This can be removed as it's not relevant to the `check` goal.

jamezp commented on src/main/resources/runners/sigtest/pom.xml:
```diff
@@ -127,22 +128,32 @@
+                <!--                  <action>strictcheck</action>-->
+                <!--                  <failOnError>true</failOnError>-->
+                <!--                  <packages>jakarta.batch.**</packages>-->
+                <!--                  <classes>${project.build.directory}/dependency</classes>-->
```
This _may_ still be needed.

jamezp commented on src/main/resources/runners/sigtest/pom.xml:
```diff
@@ -127,22 +128,32 @@
+                    </execution>
                 </executions>
+                <!--                <configuration>-->
+                <!--                  <action>strictcheck</action>-->
+                <!--                  <failOnError>true</failOnError>-->
```
I would assume these two are the defaults, but they me be required as well.

jamezp commented on src/main/resources/runners/sigtest/pom.xml:
```diff
@@ -127,22 +128,32 @@
-                  </execution>
+                    <execution>
+                        <id>default-cli</id>
+                        <!--                    <phase>integration-test</phase>-->
```
I think you still need to bind this to a goal.

liweinan commented:
Thanks for reviewing James! I'll do the modifications according to you suggestions and go on doing the tests and let you know my progress :D

liweinan commented:
@jamezp After I did the modifications according to your review comments, and then run the `sigtest` locally, I got the following errors:

```bash
➤ mvn install -Dversion.org.jberet.jberet-core=2.2.1.Final                                                                            00:01:00
...
[INFO] --- sigtest-maven-plugin:2.2:check (default-cli) @ com.ibm.jbatch.tck.sigtest.exec ---
[INFO] Packages: jakarta.batch.**
3月 23, 2024 12:01:05 上午 com.sun.tdk.signaturetest.SignatureTest parseParameters
信息: SignatureTest.args: [-FileName, /Users/weli/works/jberet-tck-runner/jakarta.batch.official.tck-2.1.1/runners/sigtest/target/sigtest-copyigtest/sigtest-1.6-batch.standalone.tck.sig-2.1-se11-OpenJDK-J9, -static, -ApiVersion, 2.2.0-SNAPSHOT, -Package, jakarta.batch, -BootCP, -Classpath, /Users/weli/works/jberet-tck-runner/jakarta.batch.official.tck-2.1.1/runners/sigtest/target/dependency:/Users/weli/works/jberet-tck-runner/jakarta.batch.official.tck-2.1.1/runners/sigtest/target/classes]
[ERROR] SignatureTest report
Base version: 2.1.0
Tested version: 2.2.0-SNAPSHOT
Check mode: src [throws normalized]
Constant checking: on

Warning: incorrect classpath parameter: /Users/weli/works/jberet-tck-runner/jakarta.batch.official.tck-2.1.1/runners/sigtest/target/classes (/Users/weli/works/jberet-tck-runner/jakarta.batch.official.tck-2.1.1/runners/sigtest/target/classes). This directory or jar file will be ignored!

Added Nested Classes
--------------------

jakarta.batch.api.partition.PartitionReducer$PartitionStatus:                   nested public final static java.lang.Enum$EnumDesc
jakarta.batch.runtime.BatchStatus:      nested public final static java.lang.Enum$EnumDesc
jakarta.batch.runtime.Metric$MetricType:nested public final static java.lang.Enum$EnumDesc

Added Superclasses or Superinterfaces
-------------------------------------

jakarta.batch.api.partition.PartitionReducer$PartitionStatus:                   interface java.lang.constant.Constable
jakarta.batch.runtime.BatchStatus:      interface java.lang.constant.Constable
jakarta.batch.runtime.Metric$MetricType:interface java.lang.constant.Constable

Added Methods
-------------

jakarta.batch.api.partition.PartitionReducer$PartitionStatus:                   method public final java.util.Optional<java.lang.Enum$EnumDesc<jakarta.batch.api.partition.PartitionReducer$PartitionStatus>> java.lang.Enum.describeConstable()
jakarta.batch.runtime.BatchStatus:      method public final java.util.Optional<java.lang.Enum$EnumDesc<jakarta.batch.runtime.BatchStatus>> java.lang.Enum.describeConstable()
jakarta.batch.runtime.Metric$MetricType:method public final java.util.Optional<java.lang.Enum$EnumDesc<jakarta.batch.runtime.Metric$MetricType>> java.lang.Enum.describeConstable()

Missed Annotations
------------------

java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=false, java.lang.String since="9")

Added Annotations
-----------------

java.lang.Enum.finalize:                anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="18")
java.lang.Enum.finalize:                anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="18")
java.lang.Enum.finalize:                anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="18")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")
java.lang.Object.finalize:              anno 0 java.lang.Deprecated(boolean forRemoval=true, java.lang.String since="9")



[INFO] /Users/weli/works/jberet-tck-runner/jakarta.batch.official.tck-2.1.1/runners/sigtest/target/surefire-reports/sigtest/TEST-com.ibm.jbatch.tck.sigtest.exec-2.2.0-SNAPSHOT.xml: 1 failures in /Users/weli/works/jberet-tck-runner/jakarta.batch.official.tck-2.1.1/runners/sigtest/target/sigtest-copy/sigtest/sigtest-1.6-batch.standalone.tck.sig-2.1-se11-OpenJDK-J9
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  3.464 s
[INFO] Finished at: 2024-03-23T00:01:05+08:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal jakarta.tck:sigtest-maven-plugin:2.2:check (default-cli) on project com.ibm.jbatch.tck.sigtest.exec: Signature check for /Users/weli/works/jberet-tck-runner/jakarta.batch.official.tck-2.1.1/runners/sigtest/target/sigtest-copy/sigtest/sigtest-1.6-batch.standalone.tck.sig-2.1-se11-OpenJDK-J9 failed with 1 -> [Help 1]
```

It seems the sig file  of `J9` cannot pass the tests under JDK 21:

<img width="1504" alt="image" src="https://github.com/jberet/jberet-tck-porting/assets/201907/35b511d8-618f-4e17-96b4-a9816a509c06">

Is there anything else I need to adjust for this to work?

jamezp commented on src/main/resources/runners/sigtest/pom.xml:
```diff
@@ -127,22 +128,38 @@
+                    <!--                        ${project.build.directory}/sigtest-copy/sigtest/sigtest-1.6-batch.standalone.tck.sig-2.1-se11-OpenJDK-J9-->
+                    <!--                    </sigfile>-->
+                    <sigfile>
+                        ${project.build.directory}/sigtest-copy/sigtest/jakarta.sigtest-2.2-batch.standalone.tck.sig-2.1-se21-Temurin
```
That path doesn't look right to me.

liweinan replied:
I guess this is related: https://github.com/jakartaee/batch-tck/pull/72#discussion_r1541492266

jamezp commented on src/main/resources/runners/sigtest/pom.xml:
```diff
@@ -127,22 +128,38 @@
+                    <action>strictcheck</action>
+                    <failOnError>true</failOnError>
+                    <packages>jakarta.batch.**</packages>
+                    <classes>${project.build.directory}/dependency</classes>
```
This will pass if you add `<release>11</release>` to let it know what the target release version is.

liweinan replied:
Cool! I'll try to verify this.

jamezp commented:
This diff works for me, but I can't open a PR for some reason https://github.com/jberet/jberet-tck-runner/compare/main...jamezp:jberet-tck-runner:sigtest-fix

liweinan commented:
Thanks for checking this James! I'll do some modifications to this according to your suggestions.

jamezp commented:
@liweinan I think #16 should work. Also is there a reason for this and the jberet-tck-runner? I would think they could be consolidated.

liweinan commented:
@jamezp Thanks for the help on this! I'll verify the commit :D 

The `jberet-tck-porting` itself contains some porting files used for the tck testings. The `jberet-tck-runner` just contains the script to run the tests.

The main reason to separate them into two projects is that the main branch of this project is also used internally for our TCK testings on Jenkins platform, which is maintained by @scottmarlow. So every change in the main branch of this project may affect the internal change.

We can merge the two projects for sure, but IMHO there are also some benefits to separate them into two projects. For example we can treat the `jberet-tck-runner` as a *consumer* to this project, and we can also maintain them in different update pace, etc.

jamezp commented:
The main benefit I see from having it one project is I could simply run:
```
mvn clean verify
```
And that would run the TCK locally for me.

liweinan commented:
After discussing with James we'll leave the current project as-is for now and improve it in the future.

liweinan commented:
replaced by: https://github.com/jberet/jberet-tck-porting/pull/16
